util/schema.go

```diff
@@ -0,0 +1,326 @@
+// operates on data tree paths.
+func IsChoiceOrCase(e *yang.Entry) bool {
+	if e == nil {
+		return false
+	}
+	return e.IsChoice() || e.IsCase()
+}
+
+// SchemaPaths returns all the paths in the path tag.
+func SchemaPaths(f reflect.StructField) ([][]string, error) {
+	var out [][]string
+	pathTag, ok := f.Tag.Lookup("path")
+	if !ok || pathTag == "" {
+		return nil, fmt.Errorf("field %s did not specify a path", f.Name)
+	}
+	if pathTag == "" {
+		return out, nil
+	}
+
+	ps := strings.Split(pathTag, "|")
+	for _, p := range ps {
+		out = append(out, StripModulePrefixes(strings.Split(p, "/")))
+	}
+	return out, nil
+}
+
+// ChildSchema returns the first child schema that matches path from the given
+// schema root. When comparing the path, only nodes that appear in the data tree
+// are considered. It returns nil if no node matches the path.
+func ChildSchema(schema *yang.Entry, path []string) *yang.Entry {
+	path = StripModulePrefixes(path)
+	entries := FindFirstNonChoiceOrCase(schema)
+
+	for _, e := range entries {
+		m := MatchingNonChoiceCaseSchema(e, path)
+		if m != nil {
+			return m
+		}
+	}
+
+	return nil
+}
+
+// FindFirstNonChoiceOrCase recursively traverses the schema tree and returns a
+// map with the set of the first nodes in every path that are neither case nor
+// choice nodes. The keys in the map are the paths to the matching elements
+// from the parent data struct, which always have length 1.
+func FindFirstNonChoiceOrCase(e *yang.Entry) map[string]*yang.Entry {
+	m := make(map[string]*yang.Entry)
+	for _, ch := range e.Dir {
+		addToEntryMap(m, findFirstNonChoiceOrCaseInternal(ch))
+	}
+	return m
+}
+
+// findFirstNonChoiceOrCaseInternal is an internal part of
+// FindFirstNonChoiceOrCase.
+func findFirstNonChoiceOrCaseInternal(e *yang.Entry) map[string]*yang.Entry {
+	m := make(map[string]*yang.Entry)
+	switch {
+	case !IsChoiceOrCase(e):
+		m[e.Name] = e
+	case e.IsDir():
+		for _, ch := range e.Dir {
+			addToEntryMap(m, findFirstNonChoiceOrCaseInternal(ch))
+		}
+	}
+	return m
+}
+
+// addToEntryMap merges from into to.
+func addToEntryMap(to, from map[string]*yang.Entry) map[string]*yang.Entry {
+	for k, v := range from {
+		to[k] = v
+	}
+	return to
+}
+
+// MatchingNonChoiceCaseSchema returns the child schema at the given path from
+// schema if one is found, or nil otherwise.
+func MatchingNonChoiceCaseSchema(schema *yang.Entry, path []string) *yang.Entry {
+	s := schema
+	if len(path) == 0 || schema.Name != path[0] {
+		return nil
+	}
+	path = path[1:]
+	for i := 0; i < len(path); i++ {
+		if s = s.Dir[path[i]]; s == nil {
+			return nil
+		}
+	}
+	return s
+}
+
+// ResolveLeafRef returns a ptr to the schema pointed to by the provided
+// leaf-ref schema. It returns schema itself if schema is not a leaf-ref.
+func ResolveLeafRef(schema *yang.Entry) (*yang.Entry, error) {
+	if schema == nil {
+		return nil, nil
+	}
+	// TODO(mostrowski): this should only be possible in fakeroot. Add an
```

Comment:
There is already an annotation in the schema for this, see [here](https://github.com/openconfig/ygot/blob/master/ygen/schemaparse.go#L58-L60).

Reply:
Done.